File: Ordina.Frontend/lib/order-online-seller-visibility.ts
/**
 * Visibilidad de pedidos para el rol "Online Seller" (equipo online, Opción A):
 * - Ver: documentos donde vendorId, referrerId o sourceReservationVendorId
 *   pertenece a cualquier usuario con rol Online Seller.
 * - Acciones (editar/eliminar/despachar): solo pedidos propios (vendorId o referrerId = usuario actual).
 */

export type OrderLikeForOnlineVisibility = {
  vendorId: string
  referrerId?: string | null
  sourceReservationVendorId?: string | null
}

function normId(value: string): string {
  return value.trim()
}

/** IDs de actores online en un pedido/reserva (vendor, referidor, vendedor de reserva origen). */
export function collectOnlineSellerActorIds(
  order: OrderLikeForOnlineVisibility,
): string[] {
  const ids: string[] = []
  const vendor = normId(order.vendorId ?? "")
  if (vendor) ids.push(vendor)
  const ref = order.referrerId?.trim()
  if (ref) ids.push(normId(ref))
  const source = order.sourceReservationVendorId?.trim()
  if (source) ids.push(normId(source))
  return ids
}

/**
 * Visible para cualquier Online Seller si algún actor del pedido es del equipo online.
 */
export function isOrderVisibleToOnlineSellerTeam(
  order: OrderLikeForOnlineVisibility,
  onlineSellerIds: ReadonlySet<string>,
): boolean {
  // El backend es la fuente de verdad; sin IDs locales no bloqueamos (p. ej. offline / cache vacío).
  if (onlineSellerIds.size === 0) return true
  return collectOnlineSellerActorIds(order).some((id) => onlineSellerIds.has(id))
}

/**
 * El pedido pertenece al usuario (puede editar/eliminar/despachar).
 */
export function isOrderOwnedByOnlineSeller(
  order: OrderLikeForOnlineVisibility,
  currentUserId: string,
): boolean {
  const uid = normId(currentUserId)
  if (!uid) return false
  if (normId(order.vendorId) === uid) return true
  const ref = order.referrerId?.trim()
  if (ref && normId(ref) === uid) return true
  return false
}

/** @deprecated Usar isOrderVisibleToOnlineSellerTeam o isOrderOwnedByOnlineSeller */
export function isOrderVisibleToOnlineSeller(
  order: OrderLikeForOnlineVisibility,
  currentUserId: string,
): boolean {
  return isOrderOwnedByOnlineSeller(order, currentUserId)
}

export function isOnlineSellerRole(role: string | undefined | null): boolean {
  if (!role?.trim()) return false
  const raw = role.trim()
  if (raw === "Online Seller") return true
  return raw.toLowerCase() === "vendedor online"
}

File: Ordina.Backend/src/Application/Orders/Ordina.Orders.Application/OnlineSeller/OrderOnlineSellerVisibility.cs
using Ordina.Database.Entities.Order;

namespace Ordina.Orders.Application.OnlineSeller;

/// <summary>
/// Reglas de visibilidad del equipo Online Seller (Opción A).
/// </summary>
public static class OrderOnlineSellerVisibility
{
    public const string OnlineSellerRole = "Online Seller";

    public static bool IsOnlineSellerRole(string? role)
    {
        if (string.IsNullOrWhiteSpace(role)) return false;
        return string.Equals(role.Trim(), OnlineSellerRole, StringComparison.Ordinal)
            || string.Equals(role.Trim(), "Vendedor Online", StringComparison.OrdinalIgnoreCase);
    }

    public static IEnumerable<string> CollectActorIds(Order order)
    {
        if (!string.IsNullOrWhiteSpace(order.VendorId))
            yield return order.VendorId.Trim();
        if (!string.IsNullOrWhiteSpace(order.ReferrerId))
            yield return order.ReferrerId.Trim();
        if (!string.IsNullOrWhiteSpace(order.SourceReservationVendorId))
            yield return order.SourceReservationVendorId.Trim();
    }

    public static bool IsVisibleToTeam(Order order, IReadOnlySet<string> onlineSellerIds)
    {
        if (onlineSellerIds.Count == 0) return false;
        return CollectActorIds(order).Any(id => onlineSellerIds.Contains(id));
    }

    public static bool IsOwnedBySeller(Order order, string userId)
    {
        if (string.IsNullOrWhiteSpace(userId)) return false;
        var uid = userId.Trim();
        if (string.Equals(order.VendorId?.Trim(), uid, StringComparison.Ordinal)) return true;
        if (!string.IsNullOrWhiteSpace(order.ReferrerId)
            && string.Equals(order.ReferrerId.Trim(), uid, StringComparison.Ordinal))
            return true;
        return false;
    }
}
